File: java/src/main/java/dsa/blind150/linkedlist/MergeTwoList.java
package dsa.blind150.linkedlist;

import dsa.utils.LoggerUtil;
import org.slf4j.Logger;

public class MergeTwoList {
    private static final Logger logger = LoggerUtil.getLogger();

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        logger.info("Starting merge process...");
        ListNode temp = new ListNode(0);
        ListNode curr = temp;

        while (l1 != null && l2 != null) {
            logger.info("Comparing values: l1 = {}, l2 = {}", l1.val, l2.val);
            if (l1.val < l2.val) {
                logger.info("Choosing l1 node with value: {}", l1.val);
                curr.next = l1;
                l1 = l1.next;
            } else {
                logger.info("Choosing l2 node with value: {}", l2.val);
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }

        // If one of the lists is not empty, link the remainder
        if (l1 != null) {
            logger.info("Remaining nodes from l1 are appended.");
        } else if (l2 != null) {
            logger.info("Remaining nodes from l2 are appended.");
        }

        curr.next = l1 == null ? l2 : l1;
        logger.info("Merge completed.");
        return temp.next;
    }

    public static void main(String[] args) {
        // Create first linked list: 1 -> 3 -> 5
        ListNode list1 = new ListNode(1);
        list1.next = new ListNode(3);
        list1.next.next = new ListNode(5);

        // Create second linked list: 2 -> 4 -> 6
        ListNode list2 = new ListNode(2);
        list2.next = new ListNode(4);
        list2.next.next = new ListNode(6);

        logger.info("Original List 1:");
        printList(list1);

        logger.info("Original List 2:");
        printList(list2);

        logger.info("Merging the two lists...");
        ListNode mergedList = mergeTwoLists(list1, list2);

        logger.info("Merged List:");
        printList(mergedList);
    }

    public static void printList(ListNode head) {
        logger.info("Printing list:");
        ListNode current = head;
        while (current != null) {
            logger.info("Node value: {}", current.val);
            current = current.next;
        }
        logger.info("End of list.");
    }
}

File: python/src/blind150/linkedList/MergeTwoList.py
from src.utils.logging_util import get_logger

logger = get_logger(__name__)
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class MergeTwoList:
    def merge_two_list(self, list1: ListNode, list2: ListNode) -> ListNode:
        temp = node = ListNode()

        while list1 and list2:
            if list1.val < list2.val:
                node.next = list1
                list1 = list1.next
            else:
                node.next = list2
                list2 = list2.next

            node = node.next

        # Linking remaining nodes from either list1 or list2
        if list1:
            node.next = list1
        elif list2:
            node.next = list2

        return temp.next

    def print_list(self, _head: ListNode):
        logger.info("Printing the list:")
        current = _head
        while current:
            logger.info(f"Node value: {current.val}")
            current = current.next
        logger.info("End of list.")

if __name__ == "__main__":
    # Creating the linked list: 1 -> 2 -> 3 -> 4 -> 5
    list1 = ListNode(1)
    list1.next = ListNode(3)
    list1.next.next = ListNode(5)

    list2 = ListNode(2)
    list2.next = ListNode(4)
    list2.next.next = ListNode(6)

    logger.info("Original Linked List 1:")
    MergeTwoList().print_list(list1)

    logger.info("Original Linked List 2:")
    MergeTwoList().print_list(list2)

    logger.info("Merging the linked list...")
    mergedList = MergeTwoList().merge_two_list(list1, list2)

    logger.info("Merged Linked List:")
    MergeTwoList().print_list(mergedList)